airflow/models.py

```diff
@@ -381,20 +392,39 @@ def kill_zombies(self, session=None):
     def bag_dag(self, dag, parent_dag, root_dag):
         """
         Adds the DAG into the bag, recurses into sub dags.
+        Throws AirflowException if there are issues with the dag. i.e. detected cycle
         """
-        self.dags[dag.dag_id] = dag
+
+        dag.test_cycle()  # throws if a task cycle is found
+
         dag.resolve_template_files()
         dag.last_loaded = timezone.utcnow()
 
         for task in dag.tasks:
             settings.policy(task)
 
-        for subdag in dag.subdags:
-            subdag.full_filepath = dag.full_filepath
-            subdag.parent_dag = dag
-            subdag.is_subdag = True
-            self.bag_dag(subdag, parent_dag=dag, root_dag=root_dag)
-        self.log.debug('Loaded DAG {dag}'.format(**locals()))
+        subdags = dag.subdags
+
+        try:
+            for subdag in subdags:
+                subdag.full_filepath = dag.full_filepath
+                subdag.parent_dag = dag
+                subdag.is_subdag = True
+                self.bag_dag(subdag, parent_dag=dag, root_dag=root_dag)
+
+            self.dags[dag.dag_id] = dag
+            self.log.debug('Loaded DAG {dag}'.format(**locals()))
+        except Exception as e:
```

Comment:
Is this a change in behavior also?

Reply:
yes the new behavior is that the bag_dag method may throw an exception if we try to bag a dag with cycles in it. Otherwise, the logic should remain the same. 

To do this, additional logic is introduced to ensure that the state of the dagbag is reverted (i.e. undo `self.dags[dag.dag_id] = dag`). This is handled by backing out of the subdag DFS and at the root dag, remove all known subdags ( `dag.subdag` actually returns all DFS subdags already). The root dag is now only added to the dagbag when all subdags are successfully added so we don't have to remove it from the dagbag manually.

Reply:
The exception is really broad, can we make this more specific?